etest/build.py

```diff
@@ -0,0 +1,154 @@
+    libc: str,
+    path: Path,
+) -> None:
+    """Build the etest image."""
+    img_profile = build_profile(quiet, architecture, libc, hardened, multilib, systemd)
+    profile = img_profile.replace("le", "")
+
+    if verbose:
+        click.echo(f"Current profile: {profile}")
+
+    build(quiet, verbose, profile, img_profile, libc, path)
+
+
+class InvalidProfileError(Exception):
+    """The profile is invalid."""
+
+    pass
+
+
+def build_profile(quiet: bool, arch: str, libc: str, hardened: bool, multilib: bool, systemd: bool):
+    """Check and create a profile string."""
+    # Sanitize the profile
+    if "arm" in arch:
+        if hardened or libc != "glibc":
+            raise InvalidProfileError("The ARM architecture can't use a different libc or hardened profiles.")
+        if systemd and arch != "arm64":
+            raise InvalidProfileError(f"{arch} doesn't support systemd.")
+    elif arch == "ppc64":
+        arch = "ppc64le"
+
+        if libc == "uclibc":
+            raise InvalidProfileError("The PPC64 architecture doesn't support uclibc.")
+        if libc == "musl":
+            raise InvalidProfileError("There are no musl base images available for PPC64.")
+        if libc == "glibc" and hardened:
+            raise InvalidProfileError("The PPC64 architecture doesn't support hardened glibc.")
+        if systemd:
+            raise InvalidProfileError("The PPC64 architecture doesn't support systemd.")
+
+    if arch != "amd64" and not multilib:
+        multilib = False
+        if not quiet:
+            click.echo("WARNING: --no-multilib is being overridden by an alternative libc.")
+
+    if libc != "glibc" and not multilib:
+        raise InvalidProfileError("Alternative libcs dont support no-multilib.")
+
+    # Build the profile string
+    profile = arch
+
+    if libc != "glibc":
+        profile += f"-{libc}"
+
+    if hardened:
+        profile += "-hardened"
+    elif libc != "glibc":
+        profile += "-vanilla"
+
+    if not multilib:
+        profile += "-nomultilib"
+
+    if systemd:
+        if profile != arch:
+            raise InvalidProfileError("Systemd profiles can't use a different libc, hardened profiles or no multilib.")
+        else:
+            profile += "-systemd"
+
+    return profile
+
+
+def build(quiet: bool, verbose: bool, profile: str, img_profile: str, libc: str, path: Path):
+    """Build the image."""
+    docker.image.build(path=str(path), buildargs={"PROFILE": img_profile}, tag=f"etest/stage1:{profile}")
+
+    if libc == "glibc":
+        stage2 = docker.container.run(
+            image=f"etest/stage1:{profile}",
+            command="/bin/bash -c \
+                    'emerge-webrsync && \
+                    echo en_US.UTF8 UTF-8 >> /etc/locale.gen && \
+                    echo en_US ISO-8859-1 >> /etc/locale.gen && \
+                    locale-gen && \
+                    eselect locale set en_US.utf8'",
+            privileged=True,
+            name="stage2",
```

Comment:
Can we use a Dockerfile that makes this simpler and not have intermediate results?

Reply:
I have had a LOT of problems with docker conditionals and webrsync requires extended privileges for x86. Its a tradeoff I had to take.

Reply:
I was thinking do the logic in Python and pass a custom Dockerfile to keep things simple.  We can go with this of course, but definitely we want to ensure that we do everything we can to ensure cleanliness in the face of failure.

Reply:
I thought of doing multiple dockerfiles, but thought it might be too messy.

Reply:
Not multiple Dockerfiles that we save (although we could do that in a docker directory), but having Python generate a Dockerfile for the given arguments and then just build that.  You'd just have to find a way to get docker to know how to work with that file (which may require a temporary directory).

Reply:
Ill look into it, alltough it seems like something that would look even more convoluted

Reply:
Yeah, I'd have to see each to know for sure.  That's part of the problem.  It smells simpler from this side though.